Add STAC <> Zarr report (#143)
cloudnativegeo/cloud-optimized-geospatial-formats-guide: cookbooks/zarr-stac-report/data-consumers/index.qmd

---
title: "Zarr + STAC for Data Consumers"
subtitle: "Accessing data cubes in Zarr + STAC"
---

> This section is targeted at people who are trying to access data cubes in Python. It addresses questions like:

Reviewer: Does this only apply to Python?

Reviewer: Maybe good to say that Python is the most elaborated but the concepts apply to other Zarr access tools ...

Author: Good point. I am quick to see the world as Python-only.

 - How do you find the data you need?
 - How do you filter down to a subset of the data?
 - When should you write to a virtualized file?

This section assumes a world in which you as a data consumer are interacting with Zarr stores cataloged by STAC.

## How to read the minimum amount of data

> One of the benefits of using a cloud-optimized file format like Zarr is that you don't have to read all the data in a dataset. You can filter the data to just the variables, area and time period that you are interested in. The sooner you do this filtering the quicker the rest of the workflow will be.

Reviewer 2: ```suggestion
One of the benefits of using a cloud-optimized file format like Zarr is that you don't have to read all the data in a dataset. Reading just the metadata to construct a logical dataset is often called lazy loading. You can filter the data to just the variables, area and time period that you are interested in. The sooner you do this filtering the quicker the rest of the workflow will be.
```

Depending on your background you might have a preference for interacting with the entire lazily-loaded data cube in xarray. Or you might want to use STAC tooling to do searching and filtering on the server-side and only start interacting with the data cube after filtering has been applied. 

> Both of those access patterns should be supported by tooling, but depending on how the catalog is set up some patterns may be simpler and faster than others (at this point in time). So which is better? Basically it depends on the structure of the Zarr store(s) and the STAC catalog that you are interacting with.

Reviewer 2: ```suggestion
Both of those access patterns are supported by tooling, but depending on how the collection is set up some patterns may be simpler and faster than others (at this point in time). So which should you use? Basically it depends on the structure of the Zarr store(s) and the STAC collection that you are interacting with. The most likely scenarios are outlined below.
```

> ## One big Zarr store in a standalone STAC collection

Reviewer 2: ```suggestion
## Scenario 1: One big Zarr store as a standalone STAC collection
```

The STAC catalog contains a collection for each Zarr store and there are collection-level assets that point to the location of the Zarr store. There are no items at all in this setup.

> In this scenario any STAC metadata exists purely for discovery and cannot be used for filtering or subsetting (see [Future Work](#future-work) for more on that). To search the STAC catalog to find collections of interest you will use the [Collection Search API Extension][7]. Depending on the level of metadata that has been provided in the STAC catalog you can search by the name of the collection and possibly by the variables (aka groups) contained within the Zarr store.

Reviewer: What property or extension would have the variables (groups)?

Author: Adding a reference to the datacube extension

### Read straight to xarray

Once you have found the collection of interest, the best approach for accessing the data is to construct the lazily-loaded data cube in xarray (or an [xarray.DataTree][8] if the Zarr store has more than one group) and filter from there.

To do this you can use the `zarr` backend directly or you can use [the `stac` backend][1] to streamline even more. The `stac` backend is mostly useful if the STAC collection uses the xarray extension.

Constructing the lazy data cube is likely to be very fast if there is a [consolidated metadata file][9] OR the data is in Zarr-3 and the Zarr metadata fetch is highly parallelized ([read more][4]).

> ## Many small Zarr stores in STAC items

Reviewer 2: ```suggestion
## Scenario 2: Many small Zarr stores as STAC items of one (1) collection
```

The STAC catalog contains many items in each collection and the items each contain one asset pointing to a Zarr store. In this setup, the Zarr store represents a particular area in time and space (a scene). This is very similar to the common practice of using COGs and STAC where each COG represents one band at one scene. The difference is instead of one band per COG, all bands are in one Zarr store (this is conceptually equivalent to a multi-band COG).

In this scenario the STAC metadata can be used to find data that overlaps the region of interest (both in time and space) before constructing a lazily-loaded data cube in xarray.

### Filter in STAC first

> If you are interacting with a STAC API and the Zarr store is cataloged in STAC following [`xstac`][10] conventions then all the metadata you need for filtering and subsetting is available at the STAC level so you should be able to search either in the web-browser directly or using pystac-client.

Reviewer: how would I know it's xstac conventions as a user?

Author: Mentioning Data Cube Extension again.

In both the API and static STAC case the result of the search will be an iterable of items. Each of these items will have at least one asset pointing to the Zarr store (sometimes there will be multiple assets that point to the same store using different protocols). Take the assets that meet your needs and you can use `xr.open_mfdataset` with the `zarr` or `stac` backend to construct the data cube. 

If the Zarr stores all cover the same area and are on the same grid then concatenation might be straightforward. If you need to apply some transformation (such as regridding or doing a projection) before concatenating the data in the files then you might find it easier to use  `xr.open_dataset` directly and iterate over the items in the list applying transformations before using one of xarray's combine methods.

If the Zarr store includes data at several resolutions (for instance if it has overviews) then it might make more sense to open it as an [xarray.DataTree][8] and pick out the arrays that are of interest before potentially reprojecting and combining across the time dimension.

> ## Virtual reference files in STAC

Reviewer 2: ```suggestion
## Scenario 3: Virtual reference files in STAC
```

> Virtual reference files can be treated exactly like how you would treat a Zarr store. They can be lazily-loaded into xarray using the [`kerchunk`][5] or [`stac`][1] backend and doing so will result in one GET request containing the minimal metadata.

Reviewer: how would a user know they have found a Virtual reference file in the STAC catalog?

Author: Adding comment about asset roles.

> Virtual reference files can be treated exactly like how you would treat a Zarr store. They can be lazily-loaded into xarray using the [`kerchunk`][5] or [`stac`][1] backend and doing so will result in one GET request containing the minimal metadata.

Reviewer 2: Should we also include icechunk here?

Author: I don't really have ideas yet about how to refer to icechunk from STAC. Maybe that can be added later on?
